graphql/core/execution/executor.py

```diff
@@ -0,0 +1,299 @@
+import collections
+import functools
+
+from ..defer import Deferred, DeferredDict, DeferredList, defer, succeed
+from ..error import GraphQLError, format_error
+from ..language import ast
+from ..language.parser import parse
+from ..language.source import Source
+from ..type import GraphQLEnumType, GraphQLInterfaceType, GraphQLList, GraphQLNonNull, GraphQLObjectType, \
+    GraphQLScalarType, GraphQLUnionType
+from ..utils import is_nullish
+from ..validation import validate
+from .base import ExecutionContext, ExecutionResult, ResolveInfo, Undefined, collect_fields, default_resolve_fn, \
+    get_argument_values, get_field_def, get_operation_root_type
+
+
+class Executor(object):
+    def __init__(self, schema, execution_middlewares=None, default_resolver=default_resolve_fn):
+        self.execution_middlewares = execution_middlewares or []
+        self.default_resolve_fn = default_resolver
+        self.schema = schema
+
+    def execute(self, request='', root=None, args=None, operation_name=None, request_context=None,
+                execute_serially=False, validate_ast=True):
+
+        curried_execution_function = functools.partial(
+            self._execute,
+            request,
+            root,
+            args,
+            operation_name,
+            request_context,
+            execute_serially,
+            validate_ast
+        )
+
+        for middleware in self.execution_middlewares:
+            if hasattr(middleware, 'execution_result'):
+                curried_execution_function = functools.partial(middleware.execution_result, curried_execution_function)
+
+        return curried_execution_function()
+
+    def _execute(self, request, root, args, operation_name, request_context, execute_serially, validate_ast):
+        if not isinstance(request, ast.Document):
+            if not isinstance(request, Source):
+                request = Source(request, 'GraphQL request')
+
+            request = parse(request)
+
+        if validate_ast:
+            validation_errors = validate(self.schema, request)
+            if validation_errors:
+                return succeed(ExecutionResult(
```

Comment:
Because graphql-js works in this way? https://github.com/graphql/graphql-js/blob/master/src/graphql.js#L40